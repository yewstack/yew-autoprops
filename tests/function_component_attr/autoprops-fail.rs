use yew::prelude::*;
use yew::html::ImplicitClone;
use yew_autoprops::*;

#[autoprops]
#[function_component]
fn CantAcceptReceiver(&self, b: bool) -> Html {
    html! {
        <p>{b}</p>
    }
}

#[autoprops]
fn not_a_function_component(b: bool) -> Html {
    html! {
        <p>{b}</p>
    }
}

#[function_component(WrongAttrsOrder)]
#[autoprops]
fn wrong_attrs_order(b: bool) -> Html {
    html! {
        <p>{b}</p>
    }
}

#[autoprops]
#[function_component(let)]
fn BadFunctionComponent(b: bool) -> Html {
    html! {
        <p>{b}</p>
    }
}

#[derive(PartialEq)]
struct NotClonable(u32);

#[autoprops]
#[function_component]
fn TypeIsNotClone(stuff: NotClonable) -> Html {
    drop(stuff);
    html! {
        <p></p>
    }
}

#[derive(Clone)]
struct NotPartialEq(u32);

impl ImplicitClone for NotPartialEq {}

#[autoprops]
#[function_component]
fn TypeIsNotPartialEq(stuff: NotPartialEq) -> Html {
    drop(stuff);
    html! {
        <p></p>
    }
}

#[autoprops]
#[function_component]
fn InvalidFieldName(_: u32) -> Html {
    html! {
        <p></p>
    }
}

mod private_module {
    #[::yew_autoprops::autoprops]
    #[::yew::function_component(CompPrivateTest)]
    fn comp_private_test(#[prop_or_default] b: ::std::primitive::bool) -> ::yew::Html
    {
        let _: ::std::primitive::bool = b;
        ::yew::html! {
            <p></p>
        }
    }
}

#[allow(unused_imports)]
use private_module::*;

fn main() {
    let _ = html! { <CompPrivateTest /> };
}
