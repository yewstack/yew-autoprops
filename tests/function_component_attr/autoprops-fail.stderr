error: function components can't accept a receiver
 --> tests/function_component_attr/autoprops-fail.rs:6:23
  |
6 | fn CantAcceptReceiver(&self, b: bool) -> Html {
  |                       ^^^^^

error: could not find #[function_component] attribute in function declaration (#[autoprops] must be placed *before* #[function_component])
  --> tests/function_component_attr/autoprops-fail.rs:13:1
   |
13 | fn not_a_function_component(b: bool) -> Html {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: expected a reference to a `Properties` type (try: `&bool`)
  --> tests/function_component_attr/autoprops-fail.rs:21:25
   |
21 | fn wrong_attrs_order(b: bool) -> Html {
   |                         ^^^^

error: expected identifier, found keyword `let`
  --> tests/function_component_attr/autoprops-fail.rs:28:22
   |
28 | #[function_component(let)]
   |                      ^^^

error: cannot use `_` as field name
  --> tests/function_component_attr/autoprops-fail.rs:61:21
   |
61 | fn InvalidFieldName(_: u32) -> Html {
   |                     ^

error[E0412]: cannot find type `CompPrivateTest` in this scope
  --> tests/function_component_attr/autoprops-fail.rs:83:22
   |
83 |     let _ = html! { <CompPrivateTest /> };
   |                      ^^^^^^^^^^^^^^^ not found in this scope
   |
note: struct `crate::private_module::CompPrivateTest` exists but is inaccessible
  --> tests/function_component_attr/autoprops-fail.rs:69:5
   |
69 |     #[::yew::function_component(CompPrivateTest)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not accessible
   = note: this error originates in the attribute macro `::yew::function_component` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NotClonable: Clone` is not satisfied
  --> tests/function_component_attr/autoprops-fail.rs:40:19
   |
38 | #[autoprops]
   | ------------ required by a bound introduced by this call
39 | #[function_component]
40 | fn TypeIsNotClone(stuff: NotClonable) -> Html {
   |                   ^^^^^ the trait `Clone` is not implemented for `NotClonable`
   |
help: consider annotating `NotClonable` with `#[derive(Clone)]`
   |
36 | #[derive(Clone)]
   |

error[E0369]: binary operation `==` cannot be applied to type `NotPartialEq`
  --> tests/function_component_attr/autoprops-fail.rs:52:30
   |
52 | fn TypeIsNotPartialEq(stuff: NotPartialEq) -> Html {
   |                              ^^^^^^^^^^^^
   |
note: an implementation of `PartialEq<_>` might be missing for `NotPartialEq`
  --> tests/function_component_attr/autoprops-fail.rs:48:1
   |
48 | struct NotPartialEq(u32);
   | ^^^^^^^^^^^^^^^^^^^ must implement `PartialEq<_>`
help: consider annotating `NotPartialEq` with `#[derive(PartialEq)]`
   |
48 | #[derive(PartialEq)]
   |
